### ISSUE TITLE
[DotNet.ProjectModel.Server] Error: An item with the same key has already been added. Key: Microsoft.CSharp

### ISSUE THREAD

**Issue by abpiskunov** [dotnet/cli#1602]
When I have a dotnet project (no sdk/version in global.json) and following project.json, I see error:
"An item with the same key has already been added. Key: Microsoft.CSharp". Error is confusing since project.json has only one dependency with such name. 

```
{
  "version": "1.0.0-*",
  "description": "KProjectB Class Library",

  "frameworks": {
    "dnx451": { 
      "dependencies": {
        "Microsoft.CSharp": "4.0.1-*",
        "System.Collections": "4.0.11-*",
        "System.Linq": "4.0.1-*",
        "System.Runtime": "4.0.21-*",
        "System.Threading": "4.0.11-*"
      }
    }
  }
}
```

**Comment by abpiskunov** [comment]
Related issue:
If immediately after I saw error above, I remove "Microsoft.CSharp" dependency from project.json (keep other packages). I see following error (warning about dependencies are expected, but error is not)


Severity	Code	Description	Project	File	Line	Suppression State
**Error		Value cannot be null.**
**Parameter name: o	KProjectLibraryNonSrc		1	**
Warning	NU1007	Dependency specified was System.Collections >= 4.0.11-* but ended up with System.Collections 4.0.0.0.	KProjectLibraryNonSrc	D:\temp\TestSolution\References\ProjectReferences\nonsrc\KProjectLibraryNonSrc\project.json	8	
Warning	NU1007	Dependency specified was System.Linq >= 4.0.1-* but ended up with System.Linq 4.0.0.0.	KProjectLibraryNonSrc	D:\temp\TestSolution\References\ProjectReferences\nonsrc\KProjectLibraryNonSrc\project.json	9	
Warning	NU1007	Dependency specified was System.Runtime >= 4.0.21-* but ended up with System.Runtime 4.0.10.0.	KProjectLibraryNonSrc	D:\temp\TestSolution\References\ProjectReferences\nonsrc\KProjectLibraryNonSrc\project.json	10	
Warning	NU1007	Dependency specified was System.Threading >= 4.0.11-* but ended up with System.Threading 4.0.0.0.	KProjectLibraryNonSrc	D:\temp\TestSolution\References\ProjectReferences\nonsrc\KProjectLibraryNonSrc\project.json	11

**Comment by troydai** [comment]
The warnings is not related in this case. If you `dotnet build` following project on command line you got a similar warning.

```
{
    "version": "1.0.0-*",
    "compilationOptions": {
        "emitEntryPoint": true
    },
    "dependencies": { },
    "frameworks": {
        "dnx451": {
            "dependencies": {
                "Microsoft.CSharp": "4.0.1-*"
            }
        }
    }
}
```

```
10:09 $ dotnet build
Compiling cli1602 for DNX,Version=v4.5.1
/Users/troydai/Code/repos/cli1602/project.json(13,37): warning NU1007: Dependency specified was Microsoft.CSharp >= 4.0.1-* but ended up with Microsoft.CSharp 4.0.0.0.

Compilation succeeded.
    1 Warning(s)
    0 Error(s)

Time elapsed 00:00:01.8385689
```

My educated guess is that the conflict is introduced by this:
https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.ProjectModel/Resolution/ProjectDependencyProvider.cs#L61

I believe the bigger problem here is that user specified “Microsoft.CSharp” doesn’t replace the package added by internal logic.

/cc @davidfowl

**Comment by abpiskunov** [comment]
See how I got second set of error + warnings. After I got duplicated CSharp dependency error, I removed it from project.json and next restore brought me a new exception: 

**Error     Value cannot be null.**
**Parameter name: o KProjectLibraryNonSrc       1  

If I reopen project there no that error anymore, so it can  be reproduced only after first error.

**Comment by troydai** [comment]
@abpiskunov let’s address the issue one by one. I’ll make a change in the https://github.com/dotnet/cli/blob/rel/1.0.0/src/Microsoft.DotNet.ProjectModel/Resolution/ProjectDependencyProvider.cs#L61 to ensure that there’s only one `Microsoft.CSharp` reference for a project reference. I’m not sure about the errors. Can you file a separate issue if you manage to build a standalone repo?

**Comment by abpiskunov** [comment]
Troy please read carefully again :). second issue reproes only after first issue and disappears after project is closed and reopened. If you just fix first issue you would never repro second one. However, the nature of that error and how it reproduces suggests that somewhere in model server code you have suspicious behavior that keeps state. See both errors as one issue please.

**Comment by troydai** [comment]
Fixed in https://github.com/dotnet/cli/commit/268dd603cce30208725f219a00506bb649a58f47

**Comment by BillHiebert** [comment]
I can still repro this with 1.0.0-beta-001888

Error		An item with the same key has already been added. Key: System.Runtime	ConsoleApp1

**Comment by troydai** [comment]
This is a different issue. The issue reported originally by @abpiskunov is specific to desktop CLR and limited to one of the four assemblies: `mscorlib`, `System`, `System.Core`, and `Microsoft.CSharp`. In this case the `System.Runtime` is a different case. @BillHiebert  can you provide the project file as well as lock file for the repo in a new issue?
